### PR TITLE
Fix shared-buffer wrapper contracts for `bindConfig()`, `bindEnv()`, and `prompt()`

### PR DESCRIPTION
**Pull request by dahlia**
Fixes https://github.com/dahlia/optique/issues/768. Part of https://github.com/dahlia/optique/issues/750.

This PR fixes inconsistent wrapper-backed dependency behavior across `object()`, `merge()`, `tuple()`, and `concat()`. Before this change, the same parser composition could resolve config and env fallbacks correctly in `object()` but fail to do so in shared-buffer constructs, and `prompt()` could still run even after a non-CLI source had already supplied the dependency value.

The implementation moves the wrapper dependency contract into shared core logic so that `bindConfig()`, `bindEnv()`, and `prompt()` all follow the same rules for state propagation, source extraction, deferred completion, and suggestion-time dependency seeding. That removes wrapper-specific drift and makes nested compositions behave the same way regardless of whether they appear in `object()`, `merge()`, `tuple()`, or `concat()`.

As part of that work, this PR also tightens annotation handling so that wrapper metadata stays available where the runtime needs it, while plain parsed values and custom parser state identities remain intact. Nested `object()` children still see parse-time annotations when they need them, invalid CLI source values remain sticky during suggestion instead of being replaced by config or env fallbacks, and cyclic plain-object values no longer trip the annotation-view unwrapping path.

From a user perspective, compositions such as `prompt(optional(bindEnv(...)))`, `prompt(withDefault(bindEnv(...)))`, and `bindConfig(...)` feeding derived parsers inside `tuple()` and `concat()` now behave consistently with the already-correct `object()` and `merge()` cases. Shared-buffer suggestion and completion paths now match actual parse behavior much more closely.

I added regression coverage across core, config, env, and inquirer tests to pin down the failure modes that motivated this change and the follow-up fixes that were needed to keep custom parser behavior stable. I also updated *CHANGES.md* and *docs/integrations/inquirer.md* to reflect the user-visible behavior changes and the corrected prompt integration behavior.

Tested with `mise test`.

### CODE REVIEW

File: packages/core/src/constructs.ts [hunk 1]
```diff
@@ -304,27 +431,150 @@ function getAnnotatedFieldState(
...
+    typeof childState !== "object" ||
+    getAnnotations(childState) === annotations
   ) {
-    return sourceState;
+    return childState;
   }
-  return Reflect.get(parser, inheritParentAnnotationsKey) === true
-    ? injectAnnotations(sourceState, annotations)
-    : inheritAnnotations(parentState, sourceState);
+  return injectAnnotations(childState, annotations);
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/constructs.ts]: Fixed in https://github.com/dahlia/optique/commit/5f752e67f1e993f43f2b95ac175933d21650e72e. `getObjectParseChildState()` now only injects annotations into plain-object child state, so custom instances keep their original identity and internal slots during parse. I also added a regression test in `constructs.test.ts`.

File: packages/core/src/constructs.ts [hunk 2]
```diff
@@ -254,11 +278,114 @@ const fieldParsersKey: unique symbol = Symbol("fieldParsers");
...
+    return source.some((item) => containsAnnotationView(item, seen));
+  }
+  const proto = Object.getPrototypeOf(source);
+  if (proto !== Object.prototype && proto !== null) {
+    return false;
+  }
+  return Reflect.ownKeys(source).some((key) =>
+    containsAnnotationView(
+      (source as Record<PropertyKey, unknown>)[key],
+      seen,
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/constructs.ts]: Fixed in https://github.com/dahlia/optique/commit/5f752e67f1e993f43f2b95ac175933d21650e72e. The annotation-view scan now walks only data-property values and skips accessors, so completion normalization no longer invokes getters. There is a matching regression test in `constructs.test.ts`.

File: packages/core/src/constructs.ts [hunk 3]
```diff
@@ -254,11 +278,114 @@ const fieldParsersKey: unique symbol = Symbol("fieldParsers");
...
+        changed = true;
+      }
+    }
+  }
+  if (!changed) {
+    seen.set(source, source);
+    return source as T;
+  }
+  Object.defineProperties(clone, descriptors);
+  return clone as T;
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/constructs.ts:377]: Fixed in https://github.com/dahlia/optique/commit/5f752e67f1e993f43f2b95ac175933d21650e72e. `withAnnotationView()` now unwraps an existing annotation view before proxying and stores the raw target in `annotationViewTargets`, so nested completion results cannot leak an inner proxy.

File: packages/env/src/index.test.ts [hunk 4]
```diff
@@ -2611,7 +2611,134 @@ describe("bindEnv() with dependency sources across merge() boundaries", () => {
...
+    assert.ok(
+      !texts.includes("debug"),
+      `Did not expect "debug" in suggestions, got: ${JSON.stringify(texts)}`,
+    );
+    assert.ok(
+      !texts.includes("verbose"),
+      `Did not expect "verbose" in suggestions, got: ${JSON.stringify(texts)}`,
+    );
+  });
+});
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/env/src/index.test.ts:2772]: Added in https://github.com/dahlia/optique/commit/6fc17d4fe0509b117fcb026d441534675ec0e5c7. The env tuple/concat suite now covers the invalid-CLI suggest case (`--mode invalid`) so failed-source precedence stays aligned with the config regressions.

File: packages/inquirer/src/index.test.ts [hunk 5]
```diff
@@ -4855,4 +4856,256 @@ describe("prompt() with dependency sources", () => {
...
+              parser,
+              ["--level", "silent"],
+              { annotations },
+            );
+
+            assert.ok(result.success);
+            assert.deepEqual(result.value, ["prod", "silent"]);
+            assert.equal(promptCalls, 0);
+          },
+        );
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/inquirer/src/index.test.ts:5013]: Added in https://github.com/dahlia/optique/commit/6fc17d4fe0509b117fcb026d441534675ec0e5c7. The inquirer env-wrapper regressions now include `suggestAsync()` coverage for both `optional()` and `withDefault()` prompt compositions.

File: packages/core/src/constructs.ts [hunk 6]
```diff
@@ -5253,6 +5575,297 @@ async function* suggestTupleAsync(
...
+              : {}),
+          };
+          matchedParsers.add(index);
+          foundMatch = true;
+          break;
+        } else if (!result.success && result.consumed < 1) {
+          matchedParsers.add(index);
+          foundMatch = true;
+          break;
+        }
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/constructs.ts:5844]: I left this logic as-is. In tuple suggestion traversal, a zero-consumption miss is intentionally treated as a skippable slot so later parsers can still inspect the same token. If this branch simply stopped advancement, optional or otherwise non-consuming tuple entries would cut off reachable completions too early. The sticky-failure handling only applies once a source parser actually consumes and fails, and I could not reproduce a concrete incorrect suggestion from the current implementation.

File: packages/core/src/constructs.ts [hunk 7]
```diff
@@ -5253,6 +5575,297 @@ async function* suggestTupleAsync(
...
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/constructs.ts:5970]: I left the async path unchanged for the same reason as the sync one. This branch only handles zero-consumption misses so later tuple parsers can still inspect the same token; consumed failures are handled separately and remain sticky. Removing it would stop async suggestion traversal too early for optional or otherwise skippable entries, and I could not reproduce a concrete incorrect completion from the current code.

File: packages/inquirer/src/index.ts [hunk 8]
```diff
@@ -739,6 +742,41 @@ export function prompt<M extends Mode, TValue, TState>(
...
+          )));
+    if (
+      cliStateIsInjectedAnnotationWrapper &&
+      requiresSourceBindingForAnnotationWrapper
+    ) {
+      return hasNestedSourceBinding;
+    }
+    return shouldAttemptInnerCompletion(cliState, state) ||
+      hasNestedSourceBinding;
+  }
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/inquirer/src/index.ts:778]: I left this unchanged after reproducing the reported cases. `prompt(optional(bindEnv(...)))` and `prompt(withDefault(bindEnv(...)))` inside both `tuple()` and `concat()` still invoke the prompter when the env value is absent, and the added sync/async regression coverage continues to pass. The singleton-array branch is still needed for resolved wrapper state, but I could not reproduce an unresolved `undefined` path that skips the prompt.

**Comment by dahlia** [comment]
Follow-up on the outside-diff review points from https://github.com/dahlia/optique/pull/769#pullrequestreview-4052244635:

- I fixed the `object()` field-state cache collision in https://github.com/dahlia/optique/commit/381b3156aca19555ed1cad5c595291bbd42a884d by namespacing the inherited child-state cache by annotator, so parse-time and completion/suggest-time annotation strategies do not reuse the same cached value.
- I left the `normalizeOptionalLikeInnerState()` opt-in condition unchanged. I re-ran the reported `optional(object(...))` / `withDefault(object(...))` nested env/config cases and could not reproduce annotation loss with the current code, while narrowing that branch to the internal opt-in marker would drop the annotation-only wrapper state that those compositions rely on.
- I also left the concat failed-source suggestion path unchanged. The current branch already preserves sticky failed-source precedence for concat suggestion flows, and the symmetric invalid-CLI tuple/concat regressions for both config and env wrappers pass on this branch.

**Comment by dahlia** [comment]
@codex review

File: packages/core/src/constructs.ts [hunk 9]
```diff
@@ -254,11 +278,118 @@ const fieldParsersKey: unique symbol = Symbol("fieldParsers");
...
+    const clone = [...source];
+    seen.set(source, clone);
+    for (let i = 0; i < source.length; i++) {
+      const nextValue = unwrapNestedAnnotationViews(source[i], seen);
+      if (nextValue !== source[i]) {
+        clone[i] = nextValue;
+        changed = true;
+      }
+    }
+    return (changed ? clone : source) as T;
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/constructs.ts]: Fixed in https://github.com/dahlia/optique/commit/0fbaefd9efa30c394d8b5cababf23f85a0ed7389. `unwrapNestedAnnotationViews()` now resets the seen entry back to the original array on the unchanged path, so shared array aliases stay stable across the whole walk.

File: packages/core/src/constructs.test.ts [hunk 10]
```diff
@@ -3214,6 +3214,293 @@ describe("object() - duplicate option detection", () => {
...
+    });
+
+    assert.ok(result.success);
+    if (result.success) {
+      assert.equal(result.value.value, "ok");
+    }
+    assert.ok(seenState != null && typeof seenState === "object");
+    assert.equal((seenState as { value: string }).value, "ok");
+    assert.ok(seenState !== childParser.initialState);
+    assert.equal(getAnnotations(seenState)?.[marker], true);
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/constructs.test.ts]: Fixed in https://github.com/dahlia/optique/commit/0fbaefd9efa30c394d8b5cababf23f85a0ed7389. The remaining `assert.equal(..., true)` checks in this coverage block were converted to `assert.ok(...)`.

File: packages/core/src/constructs.test.ts [hunk 11]
```diff
@@ -10425,6 +10787,241 @@ describe("branch coverage: constructs.ts edge cases", () => {
...
+        "tuple",
+        tuple([createCustomParser()]),
+        (value) => (value as readonly [string])[0],
+      ],
+      [
+        "concat",
+        concat(tuple([createCustomParser()])),
+        (value) => (value as readonly [string])[0],
+      ],
+    ];
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/constructs.test.ts:11059]: Fixed in https://github.com/dahlia/optique/commit/0fbaefd9efa30c394d8b5cababf23f85a0ed7389. The shared-buffer custom-child annotation matrix now includes `merge()`, and there is also a dedicated `merge()` replay regression in the same commit.

File: packages/core/src/constructs.ts [hunk 12]
```diff
@@ -5253,6 +5580,297 @@ async function* suggestTupleAsync(
   yield* deduplicateSuggestions(suggestions);
 }
 
+interface TupleSuggestAdvanceResult {
+  readonly context: ParserContext<readonly unknown[]>;
+  readonly failedParserIndexes: readonly number[];
+}
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/constructs.ts:5734]: Fixed in https://github.com/dahlia/optique/commit/622e89f06a28601accc1ed4854b25f11ab3ca51e. The tuple suggest path now only marks source IDs that actually fail in the failed subtree, instead of poisoning every source under the top-level child. I also added the nested `object()` regression that exposed the over-broad failure scope.

File: packages/core/src/constructs.ts [hunk 13]
```diff
@@ -9062,9 +9903,11 @@ export function conditional(
...
         const branchState = state.branchState ?? syncDefaultBranch.initialState;
-        const defaultResult = syncDefaultBranch.complete(
-          branchState,
-          withChildExecPath(exec, "_branch"),
+        const defaultResult = unwrapCompleteResult(
+          syncDefaultBranch.complete(
+            branchState,
+            withChildExecPath(exec, "_branch"),
+          ),
         );
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/constructs.ts]: Fixed in https://github.com/dahlia/optique/commit/0fbaefd9efa30c394d8b5cababf23f85a0ed7389. The implicit default-branch completion and the persisted conditional branch state now keep the inherited annotations, so nested source wrappers see the same context on re-entry.

File: packages/inquirer/src/index.test.ts [hunk 14]
```diff
@@ -4855,4 +4856,271 @@ describe("prompt() with dependency sources", () => {
...
+              { annotations },
+            )
+          )
+            .filter((suggestion) => suggestion.kind === "literal")
+            .map((suggestion) => suggestion.text);
+          assert.ok(suggestionTexts.includes("silent"));
+          assert.ok(suggestionTexts.includes("strict"));
+          assert.ok(!suggestionTexts.includes("debug"));
+          assert.ok(!suggestionTexts.includes("verbose"));
+        },
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/inquirer/src/index.test.ts]: Fixed in https://github.com/dahlia/optique/commit/0fbaefd9efa30c394d8b5cababf23f85a0ed7389. The remaining `bindEnv` and `bindConfig` suggest branches now assert that the prompter was not called.

File: packages/core/src/constructs.ts [hunk 15]
```diff
@@ -7154,6 +7807,7 @@ export function merge(
   // merge() does NOT forward normalizeValue because children may have
   // overlapping keys with last-write-wins semantics.  Normalizing through
   // an earlier child's normalizer would change keys owned by a later child.
+  defineInheritedAnnotationParser(mergeParser);
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/constructs.ts:7987]: Fixed in https://github.com/dahlia/optique/commit/0fbaefd9efa30c394d8b5cababf23f85a0ed7389. `merge()` now replays inherited annotations into child parse and suggest delegation the same way as the other shared-buffer constructs, and the regression coverage was extended accordingly.

File: packages/core/src/constructs.ts [hunk 16]
```diff
@@ -254,11 +278,118 @@ const fieldParsersKey: unique symbol = Symbol("fieldParsers");
...
+  const descriptors = Object.getOwnPropertyDescriptors(source) as Record<
+    PropertyKey,
+    PropertyDescriptor
+  >;
+  return Reflect.ownKeys(descriptors).some((key) => {
+    const descriptor = descriptors[key];
+    return descriptor != null && "value" in descriptor
+      ? containsAnnotationView(descriptor.value, seen)
+      : false;
+  });
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/constructs.ts:320]: I left this unchanged. That code is intentionally iterating over every own string and symbol key from the property-descriptor map while skipping accessors, so `Object.hasOwn()` is not an equivalent replacement here because this is not a single-key existence check.

**Comment by dahlia** [comment]
Follow-up on the outside-diff review at https://github.com/dahlia/optique/pull/769#pullrequestreview-4052472036:

- The `conditional()` annotation replay issue is fixed in https://github.com/dahlia/optique/commit/0fbaefd9efa30c394d8b5cababf23f85a0ed7389. Persisted branch state and implicit default-branch completion now keep the inherited annotations on re-entry.
- The tuple failed-source scoping issue is fixed in https://github.com/dahlia/optique/commit/622e89f06a28601accc1ed4854b25f11ab3ca51e. Suggest-time failure marking now only sticks to the sources that actually fail, and the nested `object()` regression that exposed the over-broad scope is covered in the same commit.
- The duplicated inline notes from that review batch about `merge()` annotation replay, the shared-buffer smoke matrix, and the boolean assertion cleanup were also handled in https://github.com/dahlia/optique/commit/0fbaefd9efa30c394d8b5cababf23f85a0ed7389.

**Comment by dahlia** [comment]
@codex review

File: packages/inquirer/src/index.ts [hunk 17]
```diff
@@ -739,6 +742,41 @@ export function prompt<M extends Mode, TValue, TState>(
...
+    const requiresSourceBindingForAnnotationWrapper =
+      Reflect.get(parser, annotationWrapperRequiresSourceBindingKey) === true;
+    const hasNestedSourceBinding = hasSourceBindingMarker(cliState) ||
+      (Array.isArray(cliState) &&
+        cliState.length === 1 &&
+        (hasSourceBindingMarker(cliState[0]) ||
+          (
+            cliState[0] != null &&
+            typeof cliState[0] === "object" &&
+            annotationKey in cliState[0]
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/inquirer/src/index.ts:768]: I left this unchanged. This branch is intentionally using `annotationKey in cliState[0]` rather than `Object.hasOwn()` because the annotation-bearing wrapper can be a proxy that exposes `annotationKey` through its `has` trap without defining it as an own property. Switching this check to `Object.hasOwn()` would miss those injected annotation wrappers and would regress the nested `prompt(optional(...))` / `prompt(withDefault(...))` source-binding path that this branch is preserving.

File: packages/core/src/constructs.ts [hunk 18]
```diff
@@ -128,14 +133,18 @@ function withChildContext<TState>(
...
   state: TState,
+  parser?: Parser<Mode, unknown, unknown>,
   usage?: Usage,
 ): ParserContext<TState> {
   const exec = withChildExecPath(context.exec, segment);
   const dependencyRegistry = context.dependencyRegistry ??
     exec?.dependencyRegistry;
+  const childState = parser == null
+    ? state
+    : getParseChildState(context.state, state, parser) as TState;
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/constructs.ts:144]: Fixed in https://github.com/dahlia/optique/commit/13b2213b3d6cc7a5544d3524b47944c2b1603fd4. `or()` and `longestMatch()` now normalize annotation-only unselected state, pass the branch parser through `withChildContext(...)`, preserve inherited annotations when storing selected branch state, and opt into parent annotation inheritance themselves. I also updated `object().parse()` to pass the child parser into `withChildContext(...)`, and added a regression test covering opt-in exclusive branches nested under `object()`.

File: packages/core/src/dependency-semantics.test.ts [hunk 19]
```diff
@@ -468,9 +464,58 @@ describe("B. Suggest path: derive() × execution contexts", () => {
...
+      "invalid",
+      "--log",
+      "",
+    ]);
+    const texts = literalTexts(suggestions);
+    assert.ok(!texts.includes("debug"), `Unexpected "debug" in ${texts}`);
+    assert.ok(!texts.includes("trace"), `Unexpected "trace" in ${texts}`);
+    assert.ok(!texts.includes("info"), `Unexpected "info" in ${texts}`);
+    assert.ok(!texts.includes("warn"), `Unexpected "warn" in ${texts}`);
+  });
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/dependency-semantics.test.ts:487]: Tightened in https://github.com/dahlia/optique/commit/6be134bd6d95f68a57cdfa708bde68da2649cb46. The invalid-source tuple regression now asserts that `literalTexts(suggestions)` is exactly empty, so any leaked literal suggestion will fail the test.

File: packages/inquirer/src/index.test.ts [hunk 20]
```diff
@@ -4855,4 +4856,250 @@ describe("prompt() with dependency sources", () => {
...
+            .filter((suggestion) => suggestion.kind === "literal")
+            .map((suggestion) => suggestion.text);
+          assert.ok(suggestionTexts.includes("silent"));
+          assert.ok(suggestionTexts.includes("strict"));
+          assert.ok(!suggestionTexts.includes("debug"));
+          assert.ok(!suggestionTexts.includes("verbose"));
+          assert.equal(promptCalls, 0);
+        },
+      );
+    }
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/inquirer/src/index.test.ts:5198]: Covered in https://github.com/dahlia/optique/commit/6be134bd6d95f68a57cdfa708bde68da2649cb46. The shared-buffer prompt suite now mirrors the `bindConfig(...)` case for both `optional(bindConfig(...))` and `withDefault(bindConfig(...))` across `tuple()` and `concat()`, so config-backed wrapper states stay pinned alongside the existing env regressions.

File: packages/core/src/constructs.test.ts [hunk 21]
```diff
@@ -3214,6 +3215,445 @@ describe("object() - duplicate option detection", () => {
...
+      ),
+      optionsTerminated: false,
+      usage: secondParser.usage,
+    });
+
+    assert.ok(secondParsed.success);
+    if (!secondParsed.success) {
+      return;
+    }
+    const secondCompleted = secondParser.complete(secondParsed.next.state);
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/constructs.test.ts:3494]: Tightened in https://github.com/dahlia/optique/commit/212d6a2dd4a26efc993a8d803cde8a1ffdb7d7d7. This regression now proves the second parse actually carries `secondMarker` before completion unwraps the nested annotation view.

File: packages/core/src/constructs.test.ts [hunk 22]
```diff
@@ -3214,6 +3215,445 @@ describe("object() - duplicate option detection", () => {
...
+    assert.ok(completeResult.success);
+    assert.ok(completeState != null && typeof completeState === "object");
+    assert.ok(getAnnotations(completeState)?.[marker] === true);
+    if (completeResult.success) {
+      assert.equal(completeResult.value.value, initialState);
+      assert.ok(
+        !Reflect.ownKeys(completeResult.value.value).includes(annotationKey),
+      );
+    }
+  });
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/constructs.test.ts:3663]: Tightened in https://github.com/dahlia/optique/commit/212d6a2dd4a26efc993a8d803cde8a1ffdb7d7d7. The cache-isolation regression now asserts that `object().complete()` receives a different child-state object than the parse-time cache.

File: packages/core/src/constructs.test.ts [hunk 23]
```diff
@@ -10356,72 +10871,391 @@ describe("branch coverage: constructs.ts edge cases", () => {
...
+    if (asyncParsed.success) {
+      const asyncCompleted = await asyncConditional.complete(
+        asyncParsed.next.state,
+      );
+      assert.ok(asyncCompleted.success);
+      if (asyncCompleted.success) {
+        assert.equal(asyncCompleted.value[0], "fast");
+      }
+    }
+  });
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/constructs.test.ts:10990]: Tightened in https://github.com/dahlia/optique/commit/212d6a2dd4a26efc993a8d803cde8a1ffdb7d7d7. The conditional completion regression now checks the retained branch payload as well as the discriminator, so re-completing from an empty optional state will fail the test.

File: packages/core/src/constructs.test.ts [hunk 24]
```diff
@@ -10356,72 +10871,391 @@ describe("branch coverage: constructs.ts edge cases", () => {
...
+        merge(
+          object({ child: createCustomParser() }),
+          object({}),
+        ),
+        (value) => (value as { readonly child: string }).child,
+      ],
+      [
+        "concat",
+        concat(tuple([createCustomParser()])),
+        (value) => (value as readonly [string])[0],
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/constructs.test.ts]: Tightened in https://github.com/dahlia/optique/commit/212d6a2dd4a26efc993a8d803cde8a1ffdb7d7d7. The new concat smoke tests now use multi-child `concat(...)` fixtures so they cross a real concat boundary instead of a single tuple child.

**Comment by dahlia** [comment]
@codex review

File: packages/core/src/constructs.test.ts [hunk 25]
```diff
@@ -10356,72 +10871,487 @@ describe("branch coverage: constructs.ts edge cases", () => {
...
+      state: {
+        discriminatorState: asyncConditional.initialState.discriminatorState,
+        discriminatorValue: undefined,
+        selectedBranch: { kind: "default" },
+        branchState: asyncDefault.initialState,
+      },
+      optionsTerminated: false,
+      usage: asyncConditional.usage,
+    });
+    assert.ok(asyncFromSelectedDefault.success);
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/constructs.test.ts:10932]: Tightened in https://github.com/dahlia/optique/commit/6783e234a26a3d245c4375602ea4114d5227a0f7. The async preselected-default case now uses a distinguishable stored default-branch state and fails from `undefined`, so it only passes when `conditional().parse()` really replays the persisted branch state.

File: packages/core/src/constructs.ts [hunk 26]
```diff
@@ -8765,7 +9794,11 @@ export function conditional(
                   discriminatorState: discriminatorResult.next.state,
                   discriminatorValue: value,
                   selectedBranch: { kind: "branch", key: value },
-                  branchState: branchParseResult.next.state,
+                  branchState: getAnnotatedChildState(
+                    state,
+                    branchParseResult.next.state,
+                    branchParser,
+                  ),
                 },
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/constructs.ts]: Fixed in https://github.com/dahlia/optique/commit/6783e234a26a3d245c4375602ea4114d5227a0f7. `conditional()` now carries inherited annotations through the discriminator parse, complete, and suggest paths as well, so wrapper-backed discriminators see the same parent annotation context as the selected branch.

**Comment by dahlia** [comment]
@codex review

File: packages/core/src/constructs.ts [hunk 27]
```diff
@@ -4412,21 +4806,35 @@ export function object<
...
+              inheritedFieldStateCache.set(parentState, nextCaches);
+              return nextCaches;
+            })();
+          return annotatorCaches.get(annotateChildState) ??
+            (() => {
+              const stateCache = new Map<string | symbol, unknown>();
+              annotatorCaches.set(annotateChildState, stateCache);
+              return stateCache;
+            })();
+        })()
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/constructs.ts:4837]: I left this as-is. This is a readability-only suggestion rather than a behavioral issue, and the current expression keeps the two-level lazy cache initialization scoped to the `parentState` object branch without introducing extra mutable locals in the hot path. Since this PR is already carrying a large bug-fix/refactor surface, I do not want to add style-only churn here.
